### ISSUE TITLE
Kubelet fails to authenticate to apiserver due to expired certificate

### ISSUE THREAD

**Issue by tshaynik** [kubernetes/kubernetes#65991]
/kind bug
/sig auth

**What happened**:
My team is having an issue with TLS bootstrap, running Kubernetes 1.10.5. We set --experimental-cluster-signing-duration to 24h on the kube-controller-manager. Some nodes are being deallocated over night, and when they come up, Kubelet goes into a failed state. It appears that it recognizes that the certificate expires and attempts to bootstrap using the token from bootstrap.kubeconfig (so far so good), but then reuses the expired certificate, and cannot authenticate to the apiserver. Here are relevant logs from kubelet:
 ```
bootstrap.go:204] Part of the existing bootstrap client certificate is expired: 2018-07-06 12:32:00 +0000 UTC                                                                                                                                                                                              
bootstrap.go:58] Using bootstrap kubeconfig to generate TLS client cert, key and kubeconfig file                                                                                                                                                                                                          
certificate_store.go:117] Loading cert/key pair from "/var/lib/kubelet/pki/kubelet-client-current.pem".                                                                                                                                                                                                    
server.go:549] Starting client certificate rotation.                                                                                                                                                                                                                                                      
certificate_manager.go:216] Certificate rotation is enabled.                                                                                                                                                                                                                                              
certificate_manager.go:287] Rotating certificates                                                                                                                                                                                                                                                          
manager.go:154] cAdvisor running in container: "/sys/fs/cgroup/cpu,cpuacct/system.slice/kubelet.service"                                                                                                                                                                                                  
certificate_manager.go:299] Failed while requesting a signed certificate from the master: cannot create certificate signing request: certificatesigningrequests.certificates.k8s.io is forbidden: User "system:anonymous" cannot create certificatesigningrequests.certificates.k8s.io at the cluster scope
```

After removing the generated cert at  `/var/lib/kubelet/pki/kubelet-client-current.pem`, kubelet was able to bootstrap properly, obtain a new cert and join the cluster.
```
rm /var/lib/kubelet/pki/kubelet-client-current.pem
systemctl restart kubelet
```
Just removing the kubeconfig, or any of the files in `/var/lib/kubelet/pki/` other than kubelet-client-current.pem and restarting kubelet did not work. Removing the entire `/var/lib/kubelet/pki/` directory and restarting kubelet works as well.

**What you expected to happen**:
I expect that after kubelet recognizes that its certificate has expired, it should remove its certificate and successfully bootstrap with the token in bootstrap.kubeconfig. It should obtain a new, valid, signed certificate from the control plane and successfully authenticate to the apiserver.

**How to reproduce it (as minimally and precisely as possible)**:
1. Set the RotateKubeletClientCertificate flag on kubelet and feature gate on kube-controller manager
1. Set the `--experimental-cluster-signing-duration` flag on the kube-controller-manager to a small duration.
1. Start kubelet with bootstrap.kubeconfig file containing a token (that is present in the token authentication file passed to the apiserver) -- kubelet bootstraps successfully and is Ready.
1. Stop kubelet before  it has attempted to renew its certificate.
1. Wait until kubelet's certificate has expired
1. Restart kubelet


**Anything else we need to know?**:
Let me know if there are more logs and information that would be useful. Thanks a lot!

**Environment**:
- Kubernetes version (use `kubectl version`): 1.10.5
- Cloud provider or hardware configuration: Azure
- OS (e.g. from /etc/os-release): CentOS 7.4
- Kernel (e.g. `uname -a`): 3.10.0-693.11.6.el7.x86_64
- Install tools: custom
- Others:

**Comment by smarterclayton** [comment]
I'm sad, because I thought this was fixed.  

/assign

**Comment by mikedanese** [comment]
At a certain point we talked about a mode for the CertificateManager where it always uses it's bootstrap kubeconfig to request a certificate. This is the type of pickle I hoped to avoid in that mode. I'm worried that the current method is too fragile to be used practically with small rotation periods.

**Comment by djsly** [comment]
Hello, we are currently setting cert rotation to be happening every 24h. 

Our users are deallocating their VM overnight to reduce the cloud cost. 

Anything we could look into as a quick fix? Or is the current recommendation *not* to use Cert Rotation until it moves out of Beta? 

thanks!

**Comment by djsly** [comment]
@mikedanese / @smarterclayton your input would be appreciated :)

**Comment by xjantoth** [comment]
I am trying to do the same:
1) In Kuberentes *v1.11.0* `rotateCertificates: true` comes out of the box:
```
kubeadm config view
...
rotateCertificates: true
...
```

2) Set *--feature-gates*
`vim vim /etc/sysconfig/kubelet`
```
KUBELET_EXTRA_ARGS=--fail-swap-on=false --feature-gates=RotateKubeletClientCertificate=true,RotateKubeletServerCertificate=true
```

3) but I how to set this flag `--experimental-cluster-signing-duration` when **kubeadm** starts?
We want to make sure that out K8s cluster will be up after one year.

Could you guys suggest what is the right aproache ?
thanks!

**Comment by mythi** [comment]
I've spent some free cycles on this and it's seems to be working OK for me on a kubeadm built cluster with kubelet version v1.11.2.

What does this mean:

> Install tools: custom

I've been using ```--experimental-cluster-signing-duration=10m``` and bootstrap tokens with different ```--ttl``` to ```kubeadm token```

**Comment by mythi** [comment]
I had just ```systemctl stop kubelet``` for 10+mins. If the bootstrap token gets expired too, bootstrapping fails but kubelet is automatically able to recover with a new token changed in my ```bootstrap-kubelet.conf```

**Comment by djsly** [comment]
ok, that's weird. maybe something new in v.1.11.2. I will @tshaynik reply since he's the one who did most of this work and investigation on our side. 

Regarding

> Install tools: custom

We install Kubernetes the hard way using a salt based solution.

**Comment by mythi** [comment]
> We install Kubernetes the hard way using a salt based solution.

OK, I'm not familiar with this.

I got almost the same error message you're seeing if I drop RBAC ```--authorization-mode``` from the api-server.

**Comment by awslovato** [comment]
> @awslovato did you manage to recover your apiserver? if so, can you share how?
> 
> I have the same situation, the apiserver won't run due to the certificate being expired and the certificate cannot be renewed due to the apiserver being down :(

Hi Albert, I did!  On the Kube Slack I found out about using kubeadm to recycle the certs that were expired with `kubeadm alpha phase certs apiserver`. I did move the old certs though, as kubeadm will not work if it sees existing certs as per: 
https://github.com/kubernetes/kubeadm/issues/649

Now my issue is that my users filled the Docker volume. I have to start all over now, and I'm not keen on that.

**Comment by joberget** [comment]
Any update on this bug? We are not using kubeadm. Only solution is to manually remove the certs.

**Comment by mosaicwang** [comment]
maybe https://github.com/kubernetes/kubeadm/issues/581#issuecomment-421477139 can solve your problem.

**Comment by liggitt** [comment]
For the original issue, 
https://github.com/kubernetes/kubernetes/pull/62152/commits/368959346af6e06085c63a4cc7c37839f262f636 was added to resolve that in 1.11

@vistalba if both the normal kubelet client certificate and the bootstrap client credential are expired, going back to your setup method to recreate a credential for the kubelet is the best recommendation

/close

**Comment by SaltedEggIndomee** [comment]
How is this resolved? I'm still experiencing this exact error on Kubernetes 1.12 running on EKS.

**Comment by awly** [comment]
@SaltedEggIndomee your error message complains about kubelet serving certificate while your workaround removes the client one.
Is that a typo in workaround steps?

In any case, serving certificate being empty and preventing kubelet startup is a separate issue, this one is about client certificates. Please file a new issue.

**Comment by gdoctor** [comment]
I can see a very similar thing happening in 1.14.x. I joined a 1.14.x node to a cluster, which completes the TLS bootstrapping process successfully, and stores current, signed certs at /var/lib/kubelet/pki/kubelet-client-current.pem.

I then delete the created /etc/kuberenetes/kubelet.conf, and join the node to a **different** 1.14.x cluster moments later. The kubelet is failing to complete the TLS bootstrapping process for the new cluster. But by also deleting /var/lib/kubelet/pki/kubelet-client-current.pem before joining the new cluster, the whole process completes as intended. 

Perhaps someone with more knowledge on this process can shed some light into the reasoning here. Btw I did not experience this with 1.13.x.
